File: 3394-minimum-array-end/minimum-array-end.cpp
class Solution {
public:
    long long minEnd(int n, int x) {
        vector<int> xbits(64,0),nbits(64,0);
        n--;
        for(int i=0;i<64;i++){
            xbits[i]=x&1;
            x>>=1;
            nbits[i]=n&1;
            n>>=1;
        }
        int i=0,j=0;
        while(i<64){
            if(!xbits[i]){
                xbits[i]=nbits[j++];
            }
            i++;
        }
        long long ans=0;
        for(int i=0;i<64;i++){
            if(xbits[i]){
                ans+=(long long)pow(2,i);
            }
        }
        return ans;
    }
};

File: 3394-minimum-array-end/minimum-array-end.java
class Solution {
    public long minEnd(int n, int x) {
        int[] xbits = new int[64];
        int[] nbits = new int[64];

        n--;

        for (int i = 0; i < 64; i++) {
            xbits[i] = x & 1;
            x >>= 1;
            nbits[i] = n & 1;
            n >>= 1;
        }

        int i = 0;
        int j = 0;
        while (i < 64) {
            if (xbits[i] == 0) {
                xbits[i] = nbits[j++];
            }
            i++;
        }

        long ans = 0;
        for (int k = 0; k < 64; k++) {
            if (xbits[k] != 0) {
                ans += Math.pow(2, k);
            }
        }

        return ans;
    }
}